server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java

```diff
@@ -87,19 +92,37 @@ public DataSegment push(File dataSegmentFile, DataSegment segment) throws IOExce
       );
     }
 
+    File tmpOutDir = new File(baseStorageDir, storageDir + "." + UUID.randomUUID().toString());
+    long size = compressSegment(dataSegmentFile, tmpOutDir);
+
+    DataSegment dataSegment = createDescriptorFile(
+      segment.withLoadSpec(makeLoadSpec(new File(outDir, "index.zip")))
+             .withSize(size)
+             .withBinaryVersion(SegmentUtils.getVersionFromDir(dataSegmentFile)),
+      tmpOutDir
+    );
+
+    // moving the temporary directory to the final destination, once success the potentially concurrent push operations
+    // will be failed and will read the descriptor.json created by current push operation directly
+    try {
+      java.nio.file.Files.move(tmpOutDir.toPath(), outDir.toPath());
+    }
+    catch (FileAlreadyExistsException e) {
+      log.warn("Push destination directory[%s] exists, ignore this message if replication is configured.");
+      dataSegment = jsonMapper.readValue(new File(outDir, "descriptor.json"), DataSegment.class);
```

Comment:
@drcrallen I think @gianm is right. I recall that our team owns one special cluster using NFS and we use a script to remove older segments.  If we upgrade the cluster to a new version with this pr, we'll have to change the script to take care of the new `intermediate_pushes` directory.  
On the other hand, if we proceed with `intermediate_pushes` directory, I think we need some explicit doc.
What do you think?